src/coins.cpp

```diff
@@ -202,10 +202,24 @@ bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn
     return true;
 }
 
-bool CCoinsViewCache::Flush() {
-    bool fOk = base->BatchWrite(cacheCoins, hashBlock);
-    cacheCoins.clear();
-    cachedCoinsUsage = 0;
+bool CCoinsViewCache::Flush(bool clear_cache) {
+    bool fOk = base->BatchWrite(cacheCoins, hashBlock, /*erase*/ clear_cache);
+    if (clear_cache) {
+        cacheCoins.clear();
+        cachedCoinsUsage = 0;
+    } else {
+        // Instead of clearing the cache, just clear the FRESH/DIRTY
+        // flags, and erase any spent coins
+        for (auto it = cacheCoins.begin(); it != cacheCoins.end(); ) {
```

Comment:
Is it true that dis loop is very slow with high dbcache? If so, a comment would be appropriate.

Other than that I also agree with @ryanofsky comment to have "separate methods like Flush() and Sync() or Flush() and Write()". As a rule of thumb, if two functions have a completely different implementation, they should not be bundled in one and switched by a boolean.

Reply:
In my own benchmarking, I haven't seen anything to support the notion that this loop takes a long time. In the newest benchmarks (to be posted), calling `Flush(erase=true)` on a cache of size 3826 MB takes **7:17** and calling `Flush(erase=false)` on a cache of the same size takes **6:03** so the difference is either negligible or favorable. 

I got the times by subtracting timestamps from the last and second-to-last loglines below for each branch during snapshot activation.

### Before this change
```
2019-12-03T23:18:26Z [snapshot] 63000000 coins loaded (99.39%, 3826 MB)
2019-12-03T23:18:26Z [snapshot] loaded 63389760 (3877 MB) coins from snapshot 00000000000000000007316856900e76b4f7a9139cfbfba89842c8d196cd5f91
2019-12-03T23:18:26Z [snapshot] flushing snapshot chainstate to disk
2019-12-03T23:25:43Z [snapshot] validated snapshot (592 MB)
```

### After this change
```
2019-12-03T23:48:09Z [snapshot] 63000000 coins loaded (99.39%, 3826 MB)
2019-12-03T23:48:10Z [snapshot] loaded 63389760 (3877 MB) coins from snapshot 00000000000000000007316856900e76b4f7a9139cfbfba89842c8d196cd5f91
2019-12-03T23:48:10Z [snapshot] flushing snapshot chainstate to disk
2019-12-03T23:54:13Z [snapshot] validated snapshot (3877 MB)
```

---

I'll make the Flush() + Sync() change.